### PR TITLE
Implement `maps_column_view` abstraction over `LIST<STRUCT<K,V>>`

### LINKED ISSUE #9109: [FEA] JNI support get map value for all value types and for all basic key types
**Is your feature request related to a problem? Please describe.**
We would like to be able to support getting map values on more than just string to string maps.  CUDF does not officially support maps and in JNI we have a stand alone special API just for string to string get map values.  This is to update that existing code unless we can add some list API operations to support it more generally.

**Describe the solution you'd like**
Update the existing JNI cudf code for get map values to be more generic. Longer term support more list operations so we can do this correctly.

**Describe alternatives you've considered**
It would be great to have a list_position. Which would return the position of the first element in a list that is equal to a value provided as either a scalar or a column vector.

We could then create map get from this by pulling out the keys from the map and calling list_position on it, and then pulling out the values list column and calling extract_list_element on it and the result of list_position.

### PR DESCRIPTION
**Pull request by mythrocks**
Fixes #9109.

This commit adds a `map` abstraction over a `column_view` of type `LIST<STRUCT<K,V>>`, where `K` and `V` are key and value types. A list column of structs with two members may thus be viewed as a `map` column. 

`maps_column_view` is to a `LIST<STRUCT<K,V>>` column what `lists_column_view` is to a `LIST` column.

The `maps_column_view` abstraction provides methods to fetch lists of keys and values (as `LIST<K>` and `LIST<V>` respectively). It also provides map lookup methods to find the values corresponding to a specified key, for each row in the "map" column.

E.g.
```c++
auto input_column = get_list_of_structs_col();
// input_column == [ {1:10, 2:20}, {1:100, 3:300}, {2:2000, 3:3000, 4:4000} ];

auto maps_view = cudf::jni::maps_column_view{input_column->view()};
auto keys = maps_view.keys();     // keys   == [ {1,2},   {1,3},      {2,3,4} ];
auto values = maps_view.values(); // values == [ {10,20}, {100, 300}, {2000, 3000, 4000} ];

auto lookup_1 = maps_view.get_values_for( *make_numeric_scalar(1) );
// lookup_1 = [ {10, 100, null} ];
```

This abstraction should help replace the Java/JNI `map_lookup` and `map_contains` kernels, which only handles `MAP<STRING, STRING>`.

### CODE REVIEW

**Comment by mythrocks** [comment]
This PR is a work in progress, and in `draft`.

**Review by jrhemstad — CHANGES_REQUESTED** [review]
Given that a map type doesn't exist in libcudf, I don't think this code should live in the libcudf header/source tree. I would expect this to live in the JNI layer or whatever we're calling the libcudf-spark companion library.

**Comment by mythrocks** [comment]
> I would expect this to live in the JNI layer or whatever we're calling the libcudf-spark companion library.

I'll move this over to the `java/src/native` directory before moving out of draft.

**Comment by mythrocks** [comment]
Note that [this file](https://github.com/mythrocks/cudf/blob/maps-column-view-impl/cpp/tests/maps/maps_tests.cpp) has tests for `maps_column_view::get_values_for()` and `maps_column_view::contains()`.

Since the JNI code doesn't currently have a space for CUDF-style tests, the test was eventually removed when `maps_column_view` was relocated to `cudf/java/src/jni`. But it has been tested with both CUDF tests and `spark-rapids` integration tests for maps.

**Comment by mythrocks** [comment]
This PR has been moved out of `Draft`, since `maps_column_view` now resides under `cudf/java`.

File: java/src/main/native/src/maps_column_view.cu
Comment (jlowe) [inline java/src/main/native/src/maps_column_view.cu:47]: Should this also check that the struct column has exactly two children?

File: java/src/main/native/src/maps_column_view.cu
Comment (jlowe) [inline java/src/main/native/src/maps_column_view.cu:62]: I think this should use`std::numeric_limits<size_type>::min()` instead.  This is similar to the gather case, where an index that is allowed to be negative can be used. Theoretically a list column could have a single row with a child column with the maximum number of elements in that single row, but the min size_type value is an invalid index even when treated as an index from the end.

File: java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java
Comment (jlowe) [inline java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java:5886]: This leaks the scalar created from `Scalar.fromInt`.

File: java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java
Comment (jlowe) [inline java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java:5843]: This leaks the scalar created from `Scalar.fromInt`.

File: java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java
Comment (jlowe) [inline java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java:5889]: This leaks the scalar created from `Scalar.fromNull`.

File: java/src/main/native/src/maps_column_view.cu
Comment (mythrocks) [reply java/src/main/native/src/maps_column_view.cu:47]: Shoot, I thought I'd added that check. Will fix.

File: java/src/main/native/src/maps_column_view.cu
Comment (mythrocks) [reply java/src/main/native/src/maps_column_view.cu:62]: > This is similar to the gather case, where an index that is allowed to be negative can be used.

It turns out that using `numeric_limits::min()` and `numeric_limit::max()` are both correct in this case.

If a list has `MAX_INT` entries, its valid (positive) indices are in the range `[0, MAX_INT - 1]`. Since `MAX_INT` falls outside that range, it is nullified.

Similarly, a list with `MAX_INT` entries would have valid negative indices in the range `[MIN_INT + 1, -1]`. Thus, `MIN_INT` would also cause `segmented_gather()` to nullify.

Would you prefer that we use `min()` instead of `max()`?

File: java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java
Comment (mythrocks) [reply java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java:5886]: Good eye. The exemplar I used had the same problem. Will fix.

File: java/src/main/native/src/maps_column_view.cu
Comment (mythrocks) [reply java/src/main/native/src/maps_column_view.cu]: Good catch. I'll remember this for the next time I need scalars of a specific type.

File: java/src/main/native/src/maps_column_view.cu
Comment (mythrocks) [reply java/src/main/native/src/maps_column_view.cu:43]: Added additional assertion.

File: java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java
Comment (mythrocks) [reply java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java]: Fixed, in all maps tests.

File: java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java
Comment (mythrocks) [reply java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java]: Fixed, in all maps tests.

File: java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java
Comment (mythrocks) [reply java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java]: Fixed, in all maps tests.

File: java/src/main/native/src/maps_column_view.cu
Comment (jlowe) [reply java/src/main/native/src/maps_column_view.cu:60]: min() would be consistent with the other cases that are using an invalid index to generate nulls.

**Review by jlowe — APPROVED** [review]
Other than the minor comment about min() vs. max(), lgtm from the Java perspective.

File: cpp/include/cudf/lists/detail/extract.hpp
Comment (ttnghia) [inline cpp/include/cudf/lists/detail/extract.hpp:33]: Why won't we have default `mr` value here?

**Review by ajschmidt8 — APPROVED** [review]
Approving `ops-codeowner` file changes

File: cpp/include/cudf/lists/detail/extract.hpp
Comment (mythrocks) [reply cpp/include/cudf/lists/detail/extract.hpp]: Good question. This was an oversight on my part. Fixed now.

File: java/src/main/native/include/maps_column_view.hpp
Comment (ttnghia) [inline java/src/main/native/include/maps_column_view.hpp:59]:
```suggestion
  lists_column_view const& keys() const { return keys_; }
```

Comment (ttnghia) [reply java/src/main/native/include/maps_column_view.hpp:59]: We should avoid copy as much as possible.

File: java/src/main/native/include/maps_column_view.hpp
Comment (mythrocks) [reply java/src/main/native/include/maps_column_view.hpp]: Good idea. Fixed.

File: java/src/main/native/include/maps_column_view.hpp
Comment (mythrocks) [reply java/src/main/native/include/maps_column_view.hpp]: Good idea. Fixed.

**Review by ttnghia — APPROVED** [review]

**Comment by mythrocks** [comment]
Thanks for the reviews, all.

I'll merge this once I have rerun tests.

**Comment by mythrocks** [comment]
@gpucibot merge

**Comment by mythrocks** [comment]
Thank you for the reviews, all. This change was just merged.

**Pull request merged**
